### PR TITLE
feat: support arbitrary uid for openshift environments

### LINKED ISSUE #125: [Feature]: support arbitrary uid for openshift environments
### Checklist

- [X] I've searched for similar issues and couldn't find anything matching
- [X] I've discussed this feature request in the [K8sGPT Slack](https://k8sgpt.slack.com) and got positive feedback

### Is this feature request related to a problem?

Yes

### Problem Description

it is the same as in https://github.com/k8sgpt-ai/k8sgpt/issues/440 which also needs to get fixed in k8sgpt-operator

### Solution Description

the operator should create a deployment which emptyDir volume and this XDG_ variable set as in https://github.com/k8sgpt-ai/k8sgpt/pull/454/commits/44725d9e84194d27b6d42504665f28ecb1f6da20

### Benefits

then on openshift k8sgpt deployment will start with restricted-v2 SCC and arbitrary uid. in addition, now k8sgpt deployment can run with securityContext.readOnlyRootFilesystem, which is also a security benefit.

### Potential Drawbacks

_No response_

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by jkleinlercher**
Closes #125

## 📑 Description
similar to https://github.com/k8sgpt-ai/k8sgpt/pull/454 k8sgpt-operator should create a deployment where XDG_ variables point to an emptyDir volume so any uid can run this container.

## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required
- [ ] All the tests have passed

### CODE REVIEW

**Comment by AlexsJones** [comment]
[ x ] Tested this does not break existing k8s clusters ( non openshift ) 
[   ] Test with OpenShift

File: pkg/resources/k8sgpt.go [hunk 1]
```diff
@@ -194,6 +194,14 @@ func GetDeployment(config v1alpha1.K8sGPT) (*appsv1.Deployment, error) {
 									Name:  "K8SGPT_BACKEND",
 									Value: string(config.Spec.Backend),
 								},
+								{
+									Name:  "XDG_CONFIG_HOME",
+									Value: "/k8sgpt-config/",
+								},
+								{
+									Name:  "XDG_CACHE_HOME",
```
Comment (arbreezy) [inline pkg/resources/k8sgpt.go:202]: I think it makes sense to have a different path for cache, e.g `/k8sgpt-config/.cache` and use subPath in the ConfigMap.

**Comment by arbreezy** [comment]
I can't run minishift in my local env and have no access to an Openshift env at the moment to test the PR.

File: pkg/resources/k8sgpt.go (hunk 1 above)
Comment (jkleinlercher) [reply pkg/resources/k8sgpt.go:202]: @arbreezy what do you think about my last namings? I use k8sgpt-data for the volume mountpath and .config and .cache sub directories

Comment (arbreezy) [reply pkg/resources/k8sgpt.go:202]: LGTM !

**Comment by jkleinlercher** [comment]
> I can't run minishift in my local env and have no access to an Openshift env at the moment to test the PR.

I can test it today on my openshift cluster

**Comment by jkleinlercher** [comment]
I build the image based on https://github.com/jkleinlercher/k8sgpt-operator/tree/feat/arbitrary-uid locally and deployed it on my openshift test cluster. I added a busybox sidecar image to the k8sgpt-deployment to check if the expected files are created in the emptyDir volume paths. As you can also see the uid of the files are the arbitery uids openshift assigns to the pod.

cache files:
```
~ $ ls -l /k8sgpt-data/.cache/k8sgpt/
total 160
-rw-------    1 10009800 10009800       452 Jun  1 07:10 010edbef1e54bd9f22a90ff8cdbd30265dccd7215454e7bb822da8d3fda2f985
-rw-------    1 10009800 10009800       612 Jun  1 07:08 036233045d065366d6a8ac063961ac1b7f89619d872af827931ddc1ee0b08306
```

config file:
```
~ $ ls -l /k8sgpt-data/.config/k8sgpt/
total 4
-rw-r--r--    1 10009800 10009800       176 Jun  1 07:02 k8sgpt.yaml
```

Also result objects in the k8sgpt namespace are created as expected

**Comment by jkleinlercher** [comment]
I tested again and now I get the following panic. Don't know if it is related to our code changes or something else currently in the main branch.

```
2023-06-01T09:42:34Z    INFO    Starting workers        {"controller": "k8sgpt", "controllerGroup": "core.k8sgpt.ai", "controllerKind": "K8sGPT", "worker count": 1}
Creating new client for 10.129.2.12:8080
Connection established between 10.129.2.12:8080 and localhost with time out of 1 seconds.
Remote Address : 10.129.2.12:8080
Local Address : 10.130.2.84:53334
2023-06-01T09:42:46Z    INFO    Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference        {"controller": "k8sgpt", "controllerGroup": "core.k8sgpt.ai", "controllerKind": "K8sGPT", "K8sGPT": {"name":"k8sgpt-sample","namespace":"sx-k8sgpt"}, "namespace": "sx-k8sgpt", "name": "k8sgpt-sample", "reconcileID": "6110b16a-bd41-4483-bcf5-dbf63d09d18c"}
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x174c11a]

goroutine 110 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.0/pkg/internal/controller/controller.go:115 +0x1fa
panic({0x190aec0, 0x2b189c0})
        /usr/local/go/src/runtime/panic.go:884 +0x213
```

**Comment by AlexsJones** [comment]
> 

Hard to know without the full stack trace, any more in the logs?

**Comment by jkleinlercher** [comment]
> Hard to know without the full stack trace, any more in the logs?

Unfortunately not, but I can say that it has nothing to do with the changes of this PR. It must be something already in main branch, because this error only occurs after the commit of merging main to this branch.

**Comment by arbreezy** [comment]
@jkleinlercher , @AlexsJones  found the issue, it because of my backstage PR, thankfully we didn't cut a release yet

this conditional will produce the error if you don't specify in a vanilla run the extra options.. we need to check first if the `extraOptions` is nil first.
https://github.com/k8sgpt-ai/k8sgpt-operator/blob/main/controllers/k8sgpt_controller.go#L224

**Review by AlexsJones — APPROVED** [review]

**Pull request merged**